README.md
# rpi-mh-z19c-exporter

Export CO₂ readings from a
[Winsen MH-Z19C](https://www.winsen-sensor.com/sensors/co2-sensor/mh-z19c.html)
CO₂ sensor connected to a Raspberry Pi to Prometheus.
The sensor needs to be connected via the UART interface.

The following metrics will be provided at the `http://localhost:1202/metrics`
endpoint:

* `co2_ppm`: Measured CO₂ concentration in parts per million (ppm).

Mnemonic for the port: 12 is C in hexadecimal and 0 = O, thus 1202 = CO2.


## Implementation

The exporter is implemented with asynchronously [Tokio](https://tokio.rs/) using
a single thread. A request to the `/metrics` endpoint will trigger reading a
new measurement from the sensor. Multiple concurrent requests will only trigger
a single measurement wait for that one to complete.


## Installation

1. Download the binary from GitHub or compile one yourself.
2. Place it on your Raspberry Pi, e.g. in `/usr/local/bin/rpi-mh-z19c-exporter`
   and ensure the the executable permission is set.
3. Use the Ansible role provided in the `roles` directory to setup a service
   user and add a systemd service. (Or do this manually if you prefer.)


## Configuration

The exporter is configured with environment variables:

* `RPI_MHZ19C_UART_PATH`: Path to UART device that the sensor is connected to.
  Default: `/dev/ttyAMA0`
* `RPI_MHZ19C_EXPORTER_LISTEN_ADDRS`: Addresses (separated by space) to listen
  on. Default: `localhost:1202`


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.2] - 2023-02-17

### Changed

* Updated dependencies


## [0.2.1] - 2022-09-24

### Fixed

* Error message strings
* Exit with non-zero exit code if worker process dies

## [0.2.0] - 2022-09-24

### Changed

* Error handling changes:
  * Print the actual error if reading the sensor fails.
  * Exit the exporter if reading the sensor fails.
  * Simplify error handling.


## [0.1.4] - 2021-10-28

### Fixed

* The program will exit when the worker dies


## [0.1.3] - 2021-04-22

### Fixed

* Links in changelog


## [0.1.2] - 2021-04-22

### Added

* Changelog

### Fixed

* Invalid crates.io keyword in `Cargo.toml`


## [0.1.1] - 2021-04-16

### Fixed

* Print correct application name at start of log.
* systemd unit name/description is now shorter


## [0.1.0] - 2021-04-16

Initial release.

[Unreleased]: https://github.com/jgosmann/rpi-mh-z19c-exporter/compare/v0.2.2...HEAD
[0.2.2]: https://github.com/jgosmann/rpi-mh-z19c-exporter/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/jgosmann/rpi-mh-z19c-exporter/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/jgosmann/rpi-mh-z19c-exporter/compare/v0.1.4...v0.2.0
[0.1.4]: https://github.com/jgosmann/rpi-mh-z19c-exporter/compare/v0.1.3...v0.1.4
[0.1.3]: https://github.com/jgosmann/rpi-mh-z19c-exporter/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/jgosmann/rpi-mh-z19c-exporter/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/jgosmann/rpi-mh-z19c-exporter/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/jgosmann/rpi-mh-z19c-exporter/releases/tag/v0.1.0



src/worker.rs
pub mod measurement_channel;

use std::{
    fmt::{self, Debug, Display, Formatter},
    time::{Duration, Instant},
};

use measurement_channel::{MeasurementReceiver, MeasurementSender};

pub type Co2MeasurementReceiver = MeasurementReceiver<u16>;
pub type Co2MeasurementSender = MeasurementSender<u16>;

pub trait Co2Sensor {
    type Error: Display;

    fn read_co2_ppm(&mut self) -> nb::Result<u16, Self::Error>;
}

pub async fn co2_sensing_worker<C: Co2Sensor>(
    mut co2_sensor: C,
    sender: MeasurementSender<u16>,
) -> C {
    loop {
        sender.notified().await;
        let value = block_with_timeout(Duration::from_millis(200), || co2_sensor.read_co2_ppm())
            .map_err(|err| match err {
                nb::Error::WouldBlock => Error::TimedOut,
                nb::Error::Other(err) => Error::SensorError(err),
            });
        match value {
            Ok(value) => {
                let send_result = sender.send_measurement(value);
                if let Err(send_error) = send_result {
                    eprintln!("internal channel send error: {}", send_error);
                    return co2_sensor;
                }
            }
            Err(err) => {
                eprintln!("sensor error: {}", err);
                return co2_sensor;
            }
        }
    }
}

fn block_with_timeout<F, T, E>(timeout: Duration, mut func: F) -> nb::Result<T, E>
where
    F: FnMut() -> nb::Result<T, E>,
{
    let abort_at = Instant::now() + timeout;
    while Instant::now() < abort_at {
        match func() {
            Err(nb::Error::WouldBlock) => (),
            result => return result,
        }
    }
    Err(nb::Error::WouldBlock)
}

#[derive(Clone, Debug)]
pub enum Error<C> {
    TimedOut,
    SensorError(C),
}

impl<C: Display> Display for Error<C> {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        use Error::*;
        match self {
            TimedOut => f.write_str("communication timeout"),
            SensorError(err) => err.fmt(f),
        }
    }
}

impl<C: Debug + Display> std::error::Error for Error<C> {}

#[cfg(test)]
pub mod tests {
    use super::*;
    use measurement_channel::measurement_channel;
    use std::collections::VecDeque;
    use std::fmt::{self, Display, Formatter};

    #[derive(Clone, Copy, Debug, Eq, PartialEq)]
    pub struct MockError;

    impl Display for MockError {
        fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), fmt::Error> {
            f.write_str("MockError")
        }
    }

    impl std::error::Error for MockError {}

    pub struct MockCo2Sensor {
        pub co2_ppm: VecDeque<nb::Result<u16, MockError>>,
    }

    impl Co2Sensor for MockCo2Sensor {
        type Error = MockError;

        fn read_co2_ppm(&mut self) -> nb::Result<u16, Self::Error> {
            self.co2_ppm
                .pop_front()
                .unwrap_or(Err(nb::Error::Other(MockError)))
        }
    }

    #[tokio::test]
    async fn test_co2_sensing_worker_normal_operation() {
        let (tx, mut rx) = measurement_channel(0u16);
        let co2_sensor = MockCo2Sensor {
            co2_ppm: VecDeque::from(vec![Ok(800)]),
        };

        tokio::spawn(async move { co2_sensing_worker(co2_sensor, tx).await });

        rx.trigger_measurement();
        assert_eq!(*rx.changed().await.unwrap(), 800);
    }
}


src/api/metrics.rs
use prometheus::proto::{Gauge, Metric, MetricFamily, MetricType};
use prometheus::Encoder;

use crate::worker::measurement_channel::MeasurementReceiver;

pub async fn serve_metrics(req: tide::Request<MeasurementReceiver<u16>>) -> tide::Result {
    req.state().trigger_measurement();
    let value = req.state().clone().changed().await?.clone();

    let mut buffer = vec![];
    let encoder = prometheus::TextEncoder::new();
    encoder.encode(&[co2_metric(value as f64)], &mut buffer)?;
    Ok(format!("{}", String::from_utf8(buffer)?).into())
}

fn co2_metric(value: f64) -> MetricFamily {
    let mut gauge = Gauge::new();
    gauge.set_value(value);

    let mut metric = Metric::new();
    metric.set_gauge(gauge);

    let mut metric_family = MetricFamily::new();
    metric_family.set_name("co2_ppm".into());
    metric_family.set_help("CO2 concentration [ppm]".into());
    metric_family.set_field_type(MetricType::GAUGE);
    metric_family.set_metric(protobuf::RepeatedField::from_slice(&[metric]));
    metric_family
}


src/measurement_channel.rs
use std::sync::Arc;
use tokio::sync::{watch, Notify};

pub fn measurement_channel<T>(initial_value: T) -> (MeasurementSender<T>, MeasurementReceiver<T>) {
    let (tx, rx) = watch::channel(initial_value);
    let notifier = Arc::new(Notify::new());
    let sender = MeasurementSender {
        notifier: notifier.clone(),
        sender: tx,
    };
    let receiver = MeasurementReceiver {
        notifier,
        receiver: rx,
    };
    (sender, receiver)
}

#[derive(Clone, Debug)]
pub struct MeasurementReceiver<T> {
    notifier: Arc<Notify>,
    receiver: watch::Receiver<T>,
}

impl<T> MeasurementReceiver<T> {
    pub fn trigger_measurement(&self) {
        self.notifier.notify_one()
    }

    pub async fn changed(&mut self) -> Result<watch::Ref<'_, T>, watch::error::RecvError> {
        self.receiver.changed().await?;
        Ok(self.receiver.borrow())
    }
}

#[derive(Debug)]
pub struct MeasurementSender<T> {
    notifier: Arc<Notify>,
    sender: watch::Sender<T>,
}

impl<T> MeasurementSender<T> {
    pub async fn notified(&self) {
        self.notifier.notified().await
    }

    pub fn send_measurement(&self, value: T) -> Result<(), watch::error::SendError<T>> {
        self.sender.send(value)
    }
}

#[cfg(test)]
pub mod tests {
    use super::*;
    use tokio::time::{timeout, Duration};

    #[tokio::test]
    async fn test_measurement_channel() {
        let timeout_duration = Duration::from_millis(10);

        let (tx, mut rx) = measurement_channel(0u16);

        rx.trigger_measurement();
        rx.trigger_measurement(); // Should only trigger one measurement
        assert!(timeout(timeout_duration, tx.notified()).await.is_ok());
        assert!(timeout(timeout_duration, tx.notified()).await.is_err());

        tx.send_measurement(42).unwrap();
        assert_eq!(
            *timeout(timeout_duration, rx.changed())
                .await
                .unwrap()
                .unwrap(),
            42
        );
    }
}


src/main.rs
mod api;
mod worker;

use crate::api::metrics::serve_metrics;
use crate::api::middleware::LogErrors;
use crate::worker::measurement_channel::measurement_channel;
use crate::worker::{co2_sensing_worker, Co2MeasurementSender, Co2Sensor};
use libsystemd::daemon::{self, NotifyState};
use mh_z19c::{self, MhZ19C};
use rppal::uart::{Parity, Uart};
use std::{convert::Into, fmt::Display};
use std::{
    fmt,
    net::{SocketAddr, ToSocketAddrs},
};
use std::{path::Path, sync::Arc};
use tide::{self};
use tokio::task::{self, JoinHandle};
use worker::Co2MeasurementReceiver;

type RpiMhZ19C<'a> = MhZ19C<'a, Uart, rppal::uart::Error>;

impl Co2Sensor for RpiMhZ19C<'_> {
    type Error = Arc<mh_z19c::Error<rppal::uart::Error>>;

    fn read_co2_ppm(&mut self) -> nb::Result<u16, Self::Error> {
        self.read_co2_ppm().map_err(|e| e.map(Arc::new))
    }
}

#[tokio::main(flavor = "current_thread")]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let config = Config::load()?;

    println!(
        "Starting rpi-mh-z19c-exporter v{} ...",
        env!("CARGO_PKG_VERSION")
    );
    print!("{}", config);

    let (tx, rx) = measurement_channel(0u16);
    let worker_process = spawn_worker(config.uart_path, tx)?;

    let app = create_app(rx);
    let server_process = app.listen(config.listen_addrs);

    println!("Ready.");
    if daemon::booted() {
        daemon::notify(true, &[NotifyState::Ready])?;
    }

    return tokio::select! {
        result = worker_process => {
            result?;
            Err("worker process died".into())
        }
        result = server_process => {
            result?;
            Ok(())
        }
    };
}

fn spawn_worker<P: AsRef<Path>>(
    uart_path: P,
    tx: Co2MeasurementSender,
) -> Result<JoinHandle<RpiMhZ19C<'static>>, Box<dyn std::error::Error>> {
    let co2sensor = MhZ19C::new(Uart::with_path(uart_path, 9600, Parity::None, 8, 1)?);
    Ok(task::spawn(async move {
        co2_sensing_worker(co2sensor, tx).await
    }))
}

fn create_app(rx: Co2MeasurementReceiver) -> tide::Server<Co2MeasurementReceiver> {
    let mut app = tide::with_state(rx);
    app.with(LogErrors);
    app.at("/metrics").get(serve_metrics);
    app
}

#[derive(Clone, Debug, PartialEq, Eq)]
struct Config {
    uart_path: String,
    listen_addrs: Vec<SocketAddr>,
}

impl Config {
    fn load() -> Result<Self, Box<dyn std::error::Error>> {
        Ok(Self {
            uart_path: std::env::var("RPI_MHZ19C_UART_PATH").unwrap_or("/dev/ttyAMA0".into()),
            listen_addrs: std::env::var("RPI_MHZ19C_EXPORTER_LISTEN_ADDRS")
                .unwrap_or("localhost:1202".into())
                .split(' ')
                .map(|addr| addr.to_socket_addrs())
                .collect::<std::io::Result<Vec<_>>>()?
                .into_iter()
                .flatten()
                .collect(),
        })
    }
}

impl Display for Config {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_fmt(format_args!("Using sensor at: {}\n", self.uart_path))?;
        f.write_str("Listening on addresses:")?;
        for addr in &self.listen_addrs {
            f.write_fmt(format_args!(" {}", addr))?;
        }
        f.write_str("\n")?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::create_app;
    use crate::worker::measurement_channel::measurement_channel;
    use crate::worker::Co2MeasurementSender;
    use tide_testing::TideTestingExt;

    struct MockWorker {
        handle: tokio::task::JoinHandle<()>,
    }

    impl MockWorker {
        fn run(tx: Co2MeasurementSender, value: u16) -> Self {
            Self {
                handle: tokio::spawn(async move {
                    loop {
                        tx.notified().await;
                        if tx.send_measurement(value).is_err() {
                            return;
                        }
                    }
                }),
            }
        }
    }

    impl Drop for MockWorker {
        fn drop(&mut self) {
            self.handle.abort();
        }
    }

    #[tokio::test]
    async fn test_serving_metrics() {
        let (tx, rx) = measurement_channel(0u16);
        let app = create_app(rx);
        let _worker = MockWorker::run(tx, 42);

        assert!(app
            .get("/metrics")
            .recv_string()
            .await
            .unwrap()
            .contains("\n# TYPE co2_ppm gauge\nco2_ppm 42\n"));
    }
}
